### PR TITLE
Add adapters

### PR DESCRIPTION
**Pull request by XuhuiZhou**
Here are some high-level points of this pull request:
* Added vision encoder in the `vision_feats` folder, allowing us to generate static image features as we want.
* Added transformer as the image adapters.
* Changed the default image embedding from ResNet to Clip, and changed the config files accordingly.
* Added code modules allowing freezing speaker's decoder and listener's encoder.
* Modified the config and bash files allowing saving models and results in different folders efficiently.

Note that we probably need to be careful about merging this branch since some of the settings may conflict what Anthony has done. The results of this pull request can be viewed [here](https://docs.google.com/spreadsheets/d/1-KQhScU-4ixeN8_GaxwkJI92z7T1h1XnFaoWU_jwrBg/edit?usp=sharing) in the `Modify Adapters and Image Embeddings` tab.

### CODE REVIEW

File: EC_finetune/__main__.py [hunk 1]
```diff
@@ -281,6 +303,22 @@ def main():
     if args.adapter_freeze_override:
         args.freeze_adapters = True
 
+    if args.sender_freeze_override:
```
Comment (shanest) [inline EC_finetune/__main__.py:306]: All of this logic seems like over-kill to me.  I think you can simplify all of this just in the `add_argument` call, by setting `dest=` and `default=` to each of those, e.g. 
```
parser.add_argument(
        '--sender_freeze_override',
        action='store_true',
        help="Flag to trigger sender freezing (overriding config)",
       dest="freeze_sender", 
       default=False
    )
```
I'm 95% about this, but check the argparse docs to confirm :)

File: EC_finetune/__main__.py [hunk 2]
```diff
@@ -289,7 +327,7 @@ def main():
         args.csv_headers = CAPTIONING_CSV_HEADERS
     else:
         args.csv_headers = EC_CSV_HEADERS
-        if args.language_model_lambda or args.weight_drift_lambda: 
+        if args.language_model_lambda or args.weight_drift_lambda:
```
Comment (shanest) [inline EC_finetune/__main__.py:330]: This feels weird; we could have an EC with neither of these right (i.e. with both lambdas at 0)? and so would still want to keep track of comm loss

File: EC_finetune/agents.py [hunk 3]
```diff
@@ -80,9 +80,20 @@ def freeze_adapters(self) -> None:
...
+            for param in self.sender.transformer.parameters():
+                param.requires_grad = False
+
+    def freeze_sender_decoder(self) -> None:
+        for param in self.sender.decoder.parameters():
+            param.requires_grad = False
+
+    def freeze_listener_encoder(self) -> None:
+        for param in self.receiver.encoder.parameters():
+            param.requires_grad = False
```
Comment (shanest) [inline EC_finetune/agents.py:96]: This can be refactored: add a little method that's like `freeze_parameters(params):` which does the for loop, and then call that from each of these three methods, just with diff args for `params` (e.g. `self.sender.lstm`, `self.sender.transformer`, etc)

File: EC_finetune/senders.py [hunk 4]
```diff
@@ -49,13 +52,15 @@ def __init__(
         self,
         model: MBartForConditionalGeneration,
         seq_len: int = None,
-        recurrent_unroll: bool = False,
+        unroll: bool = False,
```
Comment (shanest) [inline EC_finetune/senders.py:55]: Isn't it a string now, not boolean? I.e. `unroll: str = None` here

File: Vision_feats/generate_feats.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+        inference purpose only
+    """
+    def __init__(
+        self,
+        args,
+        img_name_file=None,
+    ):
+        # # TODO: Fix the place of feature extractor
+        BeitFeat = BeitFeatureExtractor.from_pretrained(
+            "microsoft/beit-large-patch16-224-pt22k"
```
Comment (shanest) [inline Vision_feats/generate_feats.py:46]: strings as args to initializer instead of hard-coded

File: Vision_feats/generate_feats.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+                    [
+                        transforms.Resize(256),
+                        transforms.CenterCrop(224),
+                        transforms.ToTensor(),
+                        transforms.Normalize(
+                            mean=[0.485, 0.456, 0.406],
+                            std=[0.229, 0.224, 0.225]
+                        ),  #Imagenet pretraining default normalization.
+                    ]
+                ),
```
Comment (shanest) [inline Vision_feats/generate_feats.py:65]: Two things here: (1) is there any way to not hard code these things for ResNet? No worries if not.  (2) this dict thing feels ad-hoc.  Maybe make a super-class with a transforms object and all of the other logic, and then subclasses for ResNet / BeIT, that basically only set the transforms to the proper thing, since that's the difference

File: Vision_feats/generate_feats.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+            'beit': MyBeit,
+            'beit_ft': MyBeit,
+            'clip': MyClip,
+            'clipL': MyClip
+        }
+        # Load model
+        self.model = self.models_dict[args.img_encode_model](args)
+        self.args = args
+        if self.args.downsample == 'transformer':
+            self.adaptor = nn.Transformer(nhead=8, num_encoder_layers=6)
```
Comment (shanest) [inline Vision_feats/generate_feats.py:197]: Less hard-coding of arguments here and below :)

File: Vision_feats/generate_feats.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Perform the forward pass with the net. Just output the feature.
+        """
+        #x = self.feat_ex(x)
+        model_output = self.model.encode_image(x).float()
+        return model_output
+
+
+class VisionEncode(nn.Module):
```
Comment (shanest) [inline Vision_feats/generate_feats.py:179]: I also feel like this might be better with subclasses.  OR minimally, making downsampler methods an argument to this initializer, instead of all of the if/else based on strings in one big downsample method

**Review by shanest — CHANGES_REQUESTED** [review]
Very nice overall; well done!! :) 

A few things, on top of the specific comments I made on the code: (1) Please coordinate with Anthony on the merge conflicts; I think those will be quick to resolve, but you two know better than me what to do.  (2) Please add a bit more detail in docstrings for classes and methods in your new vision encoder sections :).  

After those things are handled, will be very happy to merge!

File: EC_finetune/__main__.py (hunk 1 above)
Comment (XuhuiZhou) [reply EC_finetune/__main__.py:298]: I just checked the argparse docs, so we basically parse these arguments first and then we load the yml files into the args object. Then wouldn't the argument/add_argument thing above be overwrriten?

File: Configs/en-de_bt_secondary.yml [hunk 9]
```diff
@@ -30,11 +30,11 @@ early_stop_start_time: 2048
 val_dataset_script: BackTranslation/wmt19/wmt19.py
 lang_pair: de-en
 models_shared: true
-output_dir: ./Output/en-de_pipeline
+output_dir: Output/en-de_pipeline/bt_sec
```
Comment (anthonymirand) [inline Configs/en-de_bt_secondary.yml:33]: For consistency, please keep the `./Output/...` path :-)

Comment (anthonymirand) [inline Configs/en-de_bt_secondary.yml:33]: Nit: For consistency, please keep the `./Output/...` path :-)

File: Configs/en-de_captions.yml [hunk 10]
```diff
@@ -48,12 +48,12 @@ num_distractors_train: 15
...
-valid_captions: ./DataLink/captioning_new/en_captions_val.jsonl
-train_images: ./DataLink/captioning_new/images_train.pt
-valid_images: ./DataLink/captioning_new/images_val.pt
+train_captions: ./Data/captioning_new_new/en_captions_train.jsonl
+valid_captions: ./Data/captioning_new_new/en_captions_val.jsonl
+train_images: ./Data/captioning_new_new/clipL_train.pt
+valid_images: ./Data/captioning_new_new/clipL_val.pt
 save_pretrain_seperately: true
-output_dir: ./Output/en-de_pipeline/captions
+output_dir: Output/en-de_pipeline/captions
```
Comment (anthonymirand) [inline Configs/en-de_captions.yml:56]: Nit: Same here for consistent `./Output/...` path

File: Configs/en-de_bt_secondary.yml [hunk 11]
```diff
@@ -30,11 +30,11 @@ early_stop_start_time: 2048
...
 
 # data config
 # backtranslation language
-data_dir: ./DataLink/cc
+data_dir: ./Data/cc
```
Comment (anthonymirand) [inline Configs/en-de_bt_secondary.yml:37]: Should this remain `./DataLink/cc`? To my knowledge there is no `./Data/cc/en_500k.txt` uploaded to Github so this code should complain IIUC

File: Configs/en-de_captions.yml [hunk 12]
```diff
@@ -1,13 +1,13 @@
 # Model:
 model_name: ./Output/en-de_pipeline/bt_init/last
 load_entire_agent: false
-image_dim: 2048
```
Comment (anthonymirand) [inline Configs/en-de_captions.yml:4]: Are the changes to this config file (and other config files in this PR) intended to be specific to BEIT/live in a dedicated feature branch, or is this intended to be merged all into main?

If it is the latter, changing these values may impact the reproduced results of I2I. If we intend the pipeline to be run with or without BEIT (e.g. possible to run side-by-side), we may need to consider a new config to keep BEIT configs separate from the "generic" I2I configs?

File: EC_finetune/__main__.py [hunk 13]
```diff
@@ -280,16 +295,31 @@ def main():
     if args.adapter_freeze_override:
         args.freeze_adapters = True
 
+    if args.sender_freeze_override:
```
Comment (anthonymirand) [inline EC_finetune/__main__.py:298]: Nit: If I'm reading the correctly, here and below could be:

```
if args.sender_freeze_override:
  args.freeze_sender = args.sender_freeze_override:
```

Also another small readability nit, consider having the same naming conventions for the argument and override (`my_value` & `my_value_override` instead of `my_value` & `value_my_override`).

File: Configs/en-de_captions.yml [hunk 14]
```diff
@@ -48,12 +48,12 @@ num_distractors_train: 15
 num_distractors_valid: 15
 
 # Loading and saving data:
-train_captions: ./DataLink/captioning_new/en_captions_train.jsonl
-valid_captions: ./DataLink/captioning_new/en_captions_val.jsonl
-train_images: ./DataLink/captioning_new/images_train.pt
-valid_images: ./DataLink/captioning_new/images_val.pt
+train_captions: ./Data/captioning_new_new/en_captions_train.jsonl
```
Comment (anthonymirand) [inline Configs/en-de_captions.yml:51]: Similar comment as above: Should these remain `DataLink` since `Data` may be empty/are not uploaded to GitHub?

File: RunScripts/test_en-de.sh [hunk 15]
```diff
@@ -0,0 +1,5 @@
+cp /projects/unmt/communication-translation/Data/translation_references/de-en.* Output/en-de_pipeline
```
Comment (anthonymirand) [inline RunScripts/test_en-de.sh:1]: If I understand correctly, this file can be reverted since there is an existing [test_en-de_pipeline.sh](https://github.com/CLMBRs/communication-translation/blob/main/RunScripts/test_en-de_pipeline.sh) which does the same thing

File: RunScripts/en-de_pipeline.sh [hunk 16]
```diff
@@ -1,10 +1,11 @@
 #!/bin/bash
 source activate unmt
 
-OUTPUT_DIR=$1
+EX_ABBR=clipL_sFz_transf
```
Comment (anthonymirand) [inline RunScripts/en-de_pipeline.sh:4]: Should this these changes be forked in a `RunScripts/en-de_beit_pipeline.sh` to keep the existing I2I unaffected?

**Review by anthonymirand — COMMENTED** [review]
In general my comments are of the flavor: Is this feature expected to live alongside I2I/T2I? If so we may need to consider adding vision_encoder specific configs/RunScripts so I2I/T2I remains unaffected

And left some small nits for code in the existing repo (don't have relevant context for the new code)

File: EC_finetune/agents.py (hunk 3 above)
Comment (XuhuiZhou) [reply EC_finetune/agents.py:96]: good point!

File: Vision_feats/generate_feats.py (hunk 6 above)
Comment (XuhuiZhou) [reply Vision_feats/generate_feats.py:65]: (1) Good point, though I think it makes sense to hardcode these because ResNet was kind of trained on these preprocessing steps and we don't want future users to change that?
(2) Yeah, I was thinking of doing that before, somehow the transform/preprocessing step has to happen in the image loading stage since they are essential in turning the img into well-formatted tensors. (I might misunderstand what you mean here tho)

File: Vision_feats/generate_feats.py (hunk 7 above)
Comment (XuhuiZhou) [reply Vision_feats/generate_feats.py:197]: Yeah, the hard-coding thing is just for quick experiments :) So the caveat here is we actually don't use any of these downsampling methods except BEIT... and I am thinking of putting BEIT aside for now since it's not really important here. I probably would spend more time on this later if we need it. For now, does commenting out the whole downsample thing sound good to you?

File: Vision_feats/generate_feats.py (hunk 8 above)
Comment (XuhuiZhou) [reply Vision_feats/generate_feats.py:179]: See above ^ :)

File: Configs/en-de_bt_secondary.yml (hunk 11 above)
Comment (XuhuiZhou) [reply Configs/en-de_bt_secondary.yml:37]: Yeah, I guess you are right. Question: how is this DataLink thing working locally?  Do I miss any steps?

**Comment by XuhuiZhou** [comment]
> Is this feature expected to live alongside I2I/T2I? If so we may need to consider adding vision_encoder specific configs/RunScripts so I2I/T2I remains unaffected

Good point! And yes, it is. While we actually don't need extra configs/Runscripts for that since it's just one line bash thing. 
And since we still separate this vision_encoder thing in the training pipeline and just use it to generate visual embedding, I don't think that would influence I2I/T2I results.

File: RunScripts/en-de_pipeline.sh (hunk 16 above)
Comment (XuhuiZhou) [reply RunScripts/en-de_pipeline.sh:4]: Good point, I am like 98% sure this won't influence anything since that only serves as a general argument to tell the script to save the model to a different folder. I can name it `en-de_pipeline_abblation.sh` so that we don't conflict!

File: Configs/en-de_bt_secondary.yml (hunk 11 above)
Comment (XuhuiZhou) [reply Configs/en-de_bt_secondary.yml:37]: ln -s <source> DataLink

File: Configs/en-de_captions.yml (hunk 12 above)
Comment (XuhuiZhou) [reply Configs/en-de_captions.yml:4]: so basically we are not considering BEIT anymore. The `image_dim` here is changed the default to 768 because we want CLIP embeddings to be our default settings. But I am okay with just leaving 2048 as it is and probably also enabling overwritten functions so that we do not need to have separate config files for every small knob that we want to change.

**Comment by shanest** [comment]
Thank you Xuhui; looks great!!
